OpenRA.Mods.D2k/Traits/SpiceBloom.cs

```diff
@@ -0,0 +1,169 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using OpenRA.Effects;
+using OpenRA.GameRules;
+using OpenRA.Graphics;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Primitives;
+using OpenRA.Support;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.D2k.Traits
+{
+	[Desc("Seeds resources by explosive eruptions after accumulation times.")]
+	public class SpiceBloomInfo : ITraitInfo, Requires<RenderSpritesInfo>
+	{
+		[SequenceReference]
+		public readonly string[] GrowthSequences = { "grow1", "grow2", "grow3" };
+
+		[Desc("The range of time (in ticks) that the spicebloom will take to respawn.")]
+		public readonly int[] RespawnDelay = { 1500, 2500 };
+
+		[Desc("The range of time (in ticks) that the spicebloom will take to grow.")]
+		public readonly int[] GrowthDelay = { 1000, 1500 };
+
+		public readonly string ResourceType = "Spice";
+
+		[Desc("The weapon to use for spice creation.")]
+		[WeaponReference]
+		public readonly string Weapon = "SpiceExplosion";
+
+		[Desc("The amount of spice to expel.")]
+		public readonly int[] Pieces = { 3, 10 };
+
+		[Desc("The maximum distance in cells that spice may be expelled.")]
+		public readonly int Range = 5;
+
+		public object Create(ActorInitializer init) { return new SpiceBloom(init, this); }
+	}
+
+	public class SpiceBloom : ITick, INotifyKilled
+	{
+		readonly Actor self;
+		readonly SpiceBloomInfo info;
+		readonly ResourceType resType;
+		readonly ResourceLayer resLayer;
+		readonly AnimationWithOffset anim;
+
+		int counter;
+		int respawnTime;
+		int growTime;
+
+		public SpiceBloom(ActorInitializer init, SpiceBloomInfo info)
+		{
+			this.info = info;
+			self = init.Self;
+
+			resType = self.World.WorldActor.TraitsImplementing<ResourceType>().First(t => t.Info.Name == info.ResourceType);
+			resLayer = self.World.WorldActor.Trait<ResourceLayer>();
+
+			var render = self.Trait<RenderSprites>();
+			anim = new AnimationWithOffset(new Animation(init.Self.World, render.GetImage(self)), null, () => self.IsDead);
+			render.Add(anim);
+
+			respawnTime = self.World.SharedRandom.Next(info.RespawnDelay[0], info.RespawnDelay[1]);
+			growTime = self.World.SharedRandom.Next(info.GrowthDelay[0], info.GrowthDelay[1]);
+			anim.Animation.Play(info.GrowthSequences[0]);
+		}
+
+		public void Tick(Actor self)
+		{
+			if (!self.IsDead)
+			{
+				counter++;
+
+				if (counter >= growTime)
+					self.Kill(self);
+				else
+				{
+					var index = info.GrowthSequences.Length * counter / growTime;
+					anim.Animation.Play(info.GrowthSequences[index]);
+				}
+			}
+			else
+			{
+				counter++;
```

Comment:
I thought we were going to do this with aggregated damage so that attacking the spawned would accelerate the spawning instead of having two separate mechanisms that don't interact.